pennylane/templates/subroutines/out_poly.py

```diff
@@ -0,0 +1,381 @@
+        r"""Initialize the OutPoly class"""
+
+        registers_wires = [*kwargs.values(), output_wires]
+
+        num_work_wires = 0 if not work_wires else len(work_wires)
+        if mod is None:
+            mod = 2 ** len(registers_wires[-1])
+        elif mod != 2 ** len(registers_wires[-1]) and num_work_wires != 2:
+            raise ValueError(
+                f"If mod is not 2^{len(registers_wires[-1])}, two work wires should be provided"
+            )
+
+        if not isinstance(mod, int):
+            raise ValueError("mod must be integer.")
+
+        self.hyperparameters["f"] = f
+
+        all_wires = []
+        self.hyperparameters["registers_wires"] = {}
+
+        for key, value in kwargs.items():
+            wires = qml.wires.Wires(value)
+            self.hyperparameters["registers_wires"][key] = wires
+            all_wires += wires
+
+        wires = qml.wires.Wires(output_wires)
+        self.hyperparameters["registers_wires"]["output_wires"] = wires
+        all_wires += wires
+
+        self.hyperparameters["mod"] = mod
+        self.hyperparameters["work_wires"] = qml.wires.Wires(work_wires) if work_wires else None
+
+        if work_wires:
+            all_wires += work_wires
+
+        if len(all_wires) != sum(len(register) for register in registers_wires) + num_work_wires:
+            raise ValueError(
+                "None of the wires in a register must be contained in another register."
+            )
+
+        super().__init__(wires=all_wires, id=id)
+
+    def _flatten(self):
+        metadata1 = tuple(
+            (key, value) for key, value in self.hyperparameters.items() if key != "registers_wires"
+        )
+        metadata2 = tuple(self.hyperparameters["registers_wires"].items())
+
+        return tuple(), (*metadata1, *metadata2)
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+
+        hyperparams_dict = dict(metadata)
+        return cls(*data, **hyperparams_dict)
+
+    def map_wires(self, wire_map: dict):
+
+        new_registers_wires = {
+            key: qml.wires.Wires([wire_map[wire] for wire in wires])
+            for key, wires in self.hyperparameters["registers_wires"].items()
+        }
+
+        new_work_wires = (
+            [wire_map[wire] for wire in self.hyperparameters["work_wires"]]
+            if self.hyperparameters.get("work_wires")
+            else None
+        )
+
+        return OutPoly(
+            f=self.hyperparameters["f"],
+            mod=self.hyperparameters["mod"],
+            work_wires=new_work_wires,
+            **new_registers_wires,
+        )
+
+    @classmethod
+    def _primitive_bind_call(cls, *args, **kwargs):
+        return cls._primitive.bind(*args, **kwargs)
+
+    def decomposition(self):  # pylint: disable=arguments-differ
+        return self.compute_decomposition(**self.hyperparameters)
+
+    @staticmethod
+    def compute_decomposition(**kwargs):  # pylint: disable=arguments-differ
```

Comment:
Why do we use `**kwargs` here and not a similar signature to the `__init__`?

Reply:
for simplicity because everything was already included ink wargs, but I change it  👍